workflow/sync/database_semaphore.go

```diff
@@ -0,0 +1,419 @@
+				Time:       creationTime,
+			})
+		return err
+	})
+	switch err {
+	case nil:
+		s.log.WithField("key", holderKey).Debug("Added into queue")
+	default:
+		s.log.WithField("key", holderKey).WithError(err).Error("Failed to add to queue")
+	}
+}
+
+func (s *databaseSemaphore) removeFromQueue(holderKey string) {
+	_, err := s.info.session.SQL().
+		DeleteFrom(s.info.config.stateTable).
+		Where(db.Cond{stateNameField: s.dbKey}).
+		And(db.Cond{stateKeyField: holderKey}).
+		And(db.Cond{stateHeldField: false}).
+		Exec()
+	switch err {
+	case nil:
+		s.log.WithField("key", holderKey).Debug("Removed from queue")
+	default:
+		s.log.WithField("key", holderKey).WithError(err).Error("Failed to remove from queue")
+	}
+}
+
+func (s *databaseSemaphore) acquire(holderKey string) bool {
+	limit := s.getLimit()
+	result := false
+	err := s.info.session.Tx(func(sess db.Session) error {
+		existing := s.currentHoldersSession(sess)
+		if len(existing) < limit {
+			var pending []stateRecord
+			err := sess.SQL().
+				Select(stateKeyField).
+				From(s.info.config.stateTable).
+				Where(db.Cond{stateNameField: s.dbKey}).
+				And(db.Cond{stateKeyField: holderKey}).
+				And(db.Cond{stateControllerField: s.info.config.controllerName}).
+				And(db.Cond{stateMutexField: s.isMutex}).
+				And(db.Cond{stateHeldField: false}).
+				All(&pending)
+			if err != nil {
+				return err
+			}
+			if len(pending) > 0 {
+				// Update the existing row in this transaction - removeFromQueue will
+				// fail later
+				_, err := sess.SQL().Update(s.info.config.stateTable).
+					Set(stateHeldField, true).
+					Where(db.Cond{stateNameField: s.dbKey}).
+					And(db.Cond{stateKeyField: holderKey}).
+					And(db.Cond{stateControllerField: s.info.config.controllerName}).
+					And(db.Cond{stateMutexField: s.isMutex}).
+					And(db.Cond{stateHeldField: false}).
+					Exec()
+				if err != nil {
+					return err
+				}
+			} else { // insert
+				_, err := sess.Collection(s.info.config.stateTable).
+					Insert(&stateRecord{
+						Name:       s.dbKey,
+						Key:        holderKey,
+						Controller: s.info.config.controllerName,
+						Mutex:      s.isMutex,
+						Held:       true,
+					})
+				if err != nil {
+					return err
+				}
+			}
+			result = true
+			return nil
+		}
+		return nil
+	})
+	if err != nil {
+		s.log.WithField("key", holderKey).WithError(err).Error("Failed to acquire lock")
+	}
+	return result
+}
+
+// checkAcquire examines if tryAcquire would succeed
+// returns
+//
+//	true, false if we would be able to take the lock
+//	false, true if we already have the lock
+//	false, false if the lock is not acquirable
+//	string return is a user facing message when not acquirable
+func (s *databaseSemaphore) checkAcquire(holderKey string) (bool, bool, string) {
+	if holderKey == "" {
+		return false, false, "bug: attempt to check semaphore with empty holder key"
+	}
+	// Different table, no need for this to be in the Tx
+	limit := s.getLimit()
```

Comment:
I think I disagree with this assessment, it might be a different table but to be serializable this should probably be part of the transaction? 


I mean it probably doesn't matter that the limit is only eventually consistent.

Reply:
The comment could be improved.

As you say, limit only needs to be eventually consistent, if the two events are near simultaneous no-one can prove it either way.